File: Python/textbook_track/chapter06/ba6i.py
# Implement GraphToGenome
import re
import sys

from collections import namedtuple
from enum import Enum

Edge = namedtuple('Edge', ['a', 'b'])


class Sign(Enum):
    PLUS = '+'
    MINUS = '-'


class SignedNumber:
    def __init__(self, sign, number):
        self._sign = sign
        self._number = number

    def __repr__(self):
        return f'{self._sign.value}{self._number}'

    @property
    def sign(self):
        return self._sign

    @property
    def number(self):
        return self._number


class Chromosome:
    def __init__(self, synteny_blocks):
        self._synteny_blocks = tuple(synteny_blocks)

    def __repr__(self):
        return f"({' '.join(map(str, self._synteny_blocks))})"


class Cycle:
    def __init__(self, nodes):
        self._nodes = tuple(nodes)

    def __repr__(self):
        return f"({' '.join(map(str, self._nodes))})"

    def __len__(self):
        return len(self._nodes)

    def __getitem__(self, k):
        if k == len(self._nodes):
            return self._nodes[0]

        return self._nodes[k]


class Genome:
    def __init__(self, chromosomes):
        self._chromosomes = tuple(chromosomes)

    def __repr__(self):
        return ''.join(map(str, self._chromosomes))


def read_genome_graph_edges(line):
    edges = []
    for item in re.findall(r'\((.*?)\)', line):
        edge_endpoints = map(int, item.split(', '))
        edges.append(Edge(*edge_endpoints))
    return edges


def cycle_to_chromosome(cycle):
    n = len(cycle)
    synteny_blocks = []
    for ix in range(n // 2):
        node1, node2 = cycle[2*ix:2*ix+2]
        signed_number = SignedNumber(Sign.PLUS, node2 // 2) if node1 < node2 else SignedNumber(Sign.MINUS, node1 // 2)
        synteny_blocks.append(signed_number)
    return Chromosome(synteny_blocks)


def identify_cycles(colored_edges):
    cycles = []
    current_cycle = []
    for edge in colored_edges:
        if not current_cycle:
            start_node = edge.a

        current_cycle.extend(edge)
        potential_last = edge.b - 1 if edge.b % 2 == 0 else edge.b + 1
        if potential_last == start_node:
            end_node = current_cycle.pop(-1)
            cycles.append(Cycle([end_node] + current_cycle))
            current_cycle = []

    return cycles


def graph_to_genome(colored_edges):
    cycles = identify_cycles(colored_edges)
    chromosomes = map(cycle_to_chromosome, cycles)
    return Genome(chromosomes)


def main():
    reader = sys.stdin
    colored_edges = read_genome_graph_edges(next(reader))
    result = graph_to_genome(colored_edges)
    print(result)


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter06/BA6I.scala
package TextbookTrack.Chapter06

object BA6I {
  import scala.annotation.tailrec
  import scala.util.matching.Regex

  sealed trait Sign
  final case object Plus extends Sign {
    override def toString: String = "+"
  }
  final case object Minus extends Sign {
    override def toString: String = "-"
  }

  object Sign {
    def apply(char: Char): Sign = char match {
      case '+' => Plus
      case '-' => Minus
      case _ => throw new Exception("Unknown sign.")
    }
  }

  final case class SignedNumber(sign: Sign, number: Int) {
    override def toString: String = s"$sign$number"
  }

  final case class Chromosome(syntenyBlocks: Vector[SignedNumber]) {
    override def toString: String = syntenyBlocks.mkString("(", " ", ")")
  }

  final case class Cycle(nodes: Vector[Int]) {
    override def toString: String = nodes.mkString("(", " ", ")")

    def grouped(size: Int): Iterator[Vector[Int]] = nodes.grouped(size)
  }

  final case class Genome(chromosomes: List[Chromosome]) {
    override def toString: String = chromosomes.map(_.toString).mkString
  }

  final case class Edge(a: Int, b: Int) {
    override def toString: String = s"($a, $b)"
  }

  private def readGenomeGraphEdges(line: String): List[Edge] = {
    val pattern: Regex = "\\((.*?)\\)".r
    pattern
      .findAllIn(line)
      .map(_.dropRight(1).drop(1).split(", ").map(_.toInt).toList)
      .collect{ case List(a, b) => Edge(a, b) }
      .toList
  }

  private def cycleToChromosome(cycle: Cycle): Chromosome = {
    val syntenyBlocks: Vector[SignedNumber] =
      cycle
        .grouped(2)
        .collect{
          case Vector(node1, node2) =>
            if (node1 < node2) SignedNumber(Plus, node2 / 2)
            else SignedNumber(Minus, node1 / 2)
        }
        .toVector
    Chromosome(syntenyBlocks)
  }

  private def identifyCycles(coloredEdges: List[Edge]): List[Cycle] = {
    @tailrec
    def loop(cycles: List[Cycle],
             currentNodes: List[Int],
             startNode: Option[Int],
             edges: List[Edge]): List[Cycle] = edges match {
      case Nil => cycles.reverse
      case Edge(a, b) :: rest =>
        val start: Int = startNode match {
          case None => a
          case Some(node) => node
        }
        val potentialLast: Int = if (b % 2 == 0) b - 1 else b + 1
        if (potentialLast == start) {
          val nodes: List[Int] = b :: (a :: currentNodes).reverse
          val cycle: Cycle = Cycle(nodes.toVector)
          loop(cycle :: cycles, Nil, None, rest)
        } else loop(cycles, b :: a :: currentNodes, Some(start), rest)
    }

    loop(Nil, Nil, None, coloredEdges)
  }

  def graphToGenome(coloredEdges: List[Edge]): Genome = {
    val cycles: List[Cycle] = identifyCycles(coloredEdges)
    val chromosomes: List[Chromosome] = cycles.map(cycleToChromosome)
    Genome(chromosomes)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val coloredEdges: List[Edge] = readGenomeGraphEdges(reader.next())
    val result: Genome = graphToGenome(coloredEdges)
    println(result)
  }
}
